ingress: rename backend to defaultBackend for ingress v1

cmluciano opened the pull request:
Signed-off-by: Christopher M. Luciano <cmluciano@us.ibm.com>

**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
Per the Ingress v1 KEP, we are renaming the backend field in IngressSpec to
clarify its intended use. Appropriate defaults are added to ensure backend is
converted to the new field type.

**Which issue(s) this PR fixes**:
Related kubernetes/enhancements#758

**Special notes for your reviewer**:
- [x] Update APIs
- [x] Add conversion logic for v1beta1 & v1
- [x] Add conversion logic tests

**Does this PR introduce a user-facing change?**:

```release-note
ingress: rename backend to defaultBackend for ingress v1
```

cmluciano commented:
cc @aledbf @bowei

bowei commented:
thanks for picking this up :-)

cmluciano commented:
/priority important-soon
/milestone 1.16

cmluciano commented:
Just need to add tests tomorrow and then will flip this to "ready"

cmluciano commented:
@kubernetes/api-approvers Can anyone spot the obvious errors in my conversion code for the v1beta1 apis? It seems to be fine in v1.

liggitt commented on pkg/apis/networking/types.go:
```diff
@@ -239,9 +239,19 @@ type IngressSpec struct {
+	// and will replace the Backend type entirely in a future release.
+	// This type is GA for 1.16
+	// +optional
+	DefaultBackend *IngressBackend
```
the internal type should only have a single field. the `backend` field in v1beta1 and the `defaultBackend` field in v1 should both convert to/from the same internal field

liggitt replied:
if you name the internal field to `DefaultBackend`, you will need to write v1beta1 custom conversion functions to convert the v1beta1 `Backend` field to/from that field.

if you name the internal field to `Backend`, you will need to write v1 custom conversion functions to convert the v1 `DefaultBackend` field to/from that field.

cmluciano replied:
That makes more sense, thanks a ton @liggitt

liggitt commented on staging/src/k8s.io/api/networking/v1/types.go:
```diff
@@ -201,3 +201,174 @@ type NetworkPolicyList struct {
+	// The default backend capable of servicing requests that don't match any
+	// rule. At least one of 'defaultBackend' or 'rules' must be specified. This field
+	// is optional to allow the loadbalancer controller or defaulting logic to
+	// specify a global default. This type is a rename for the spec.Backend for clarity
```
drop the rename/ga comments here

cmluciano commented:
/label api-review

liggitt commented:
this is introducing the v1 types. is this the full list of changes being made in v1, or are there others as well? we want to make sure we don't ship the v1 API with changes half-done

liggitt commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -502,3 +504,45 @@ func Convert_networking_NetworkPolicyList_To_v1beta1_NetworkPolicyList(in *netwo
 	return nil
 }
+
+func Convert_v1beta1_IngressSpec_To_networking_IngressSpec(in *extensionsv1beta1.IngressSpec, out *networking.IngressSpec, s conversion.Scope) error {
```
prefer calling autoConvert_v1beta1_IngressSpec_To_networking_IngressSpec for all the fields that could be converted automatically, then just convert the backend/defaultBackend field manually

liggitt commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -502,3 +504,45 @@ func Convert_networking_NetworkPolicyList_To_v1beta1_NetworkPolicyList(in *netwo
+	return nil
+}
+
+func Convert_networking_IngressSpec_To_v1beta1_IngressSpec(in *networking.IngressSpec, out *extensionsv1beta1.IngressSpec, s conversion.Scope) error {
```
same here for autoConvert_networking_IngressSpec_To_v1beta1_IngressSpec

liggitt commented on pkg/apis/networking/types.go:
```diff
@@ -235,12 +235,14 @@ type IngressList struct {
-	// specify a global default.
+	// specify a global default. This type is a rename for the spec.Backend for clarity
+	// and will replace the Backend type entirely in a future release.
+	// This type is GA for 1.16
```
drop "This type is GA for 1.16", that applies to the external v1 type and is implicit in the version

liggitt commented on pkg/apis/networking/v1beta1/conversion.go:
```diff
@@ -0,0 +1,73 @@
+	)
+}
+
+func Convert_v1beta1_IngressSpec_To_networking_IngressSpec(in *v1beta1.IngressSpec, out *networking.IngressSpec, s conversion.Scope) error {
```
same comment about autoConvert_v1beta1_IngressSpec_To_networking_IngressSpec

thockin replied:
The original code here probably pre-dates the `autoConvert` trick, but I agree.  Call autoConvert and then fix up the things that it doesn't handle.

liggitt commented on pkg/apis/networking/v1beta1/conversion.go:
```diff
@@ -0,0 +1,73 @@
+	return nil
+}
+
+func Convert_networking_IngressSpec_To_v1beta1_IngressSpec(in *networking.IngressSpec, out *v1beta1.IngressSpec, s conversion.Scope) error {
```
same comment about autoConvert_networking_IngressSpec_To_v1beta1_IngressSpec

liggitt commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -502,3 +504,45 @@ func Convert_networking_NetworkPolicyList_To_v1beta1_NetworkPolicyList(in *netwo
+
+func Convert_v1beta1_IngressSpec_To_networking_IngressSpec(in *extensionsv1beta1.IngressSpec, out *networking.IngressSpec, s conversion.Scope) error {
+	if in.Backend != nil {
+		if err := s.Convert(&in.Backend, &out.DefaultBackend, 0); err != nil {
```
this can silently rely on reflective conversion, which is very slow. use `Convert_v1beta1_IngressBackend_To_networking_IngressBackend` instead (similar comment applies to all other uses of s.Convert in the new functions)

liggitt commented on pkg/apis/networking/v1beta1/conversion_test.go:
```diff
@@ -0,0 +1,60 @@
+	"testing"
+)
+
+func TestNilOrEmptyConversion(t *testing.T) {
```
how is this testing nil or empty conversion?

liggitt commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -209,10 +209,10 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
-	if spec.Backend != nil {
-		allErrs = append(allErrs, validateIngressBackend(spec.Backend, fldPath.Child("backend"))...)
+	if spec.DefaultBackend != nil {
+		allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child("defaultBackend"))...)
```
this will produce confusing validation messages for existing v1beta1 API callers... since the `defaultBackend` field doesn't exist in their API versions. for this to be coherent, this needs to know either what version is being validated, or the field name to use for the backend field

thockin replied:
This is a fundamental problem with validation today - it always runs on the INTERNAL type, which means error messages in this situation are confusing.

I don't know of a better solution.  https://github.com/kubernetes/kubernetes/issues/8550

Suggestions welcome.

thockin replied:
I spoke to @lavalamp about versioned validation.  He said he'd support an approach to start moving to versioned validation, but wasn't sure exactly what the right hook is.  If we want to start exploring that in parallel, I'm happy to try to help try to figure it out.

In the mean time, I think we should go with the new name in errors.

The alternative is to plumb operated-as-version strings down into validation (and updates).  Blech.

liggitt replied:
> In the mean time, I think we should go with the new name in errors.
> 
> The alternative is to plumb operated-as-version strings down into validation (and updates). Blech.

as a user, it is maddening to get an API error, follow the instructions, and not resolve the error (or produce a different "unknown field" error). I'd probably plumb a map of internal to external name for this one until the per-version validation approach materializes

cmluciano replied:
So should we switch the err to include the old spec field in there or do we want to go more complex?

thockin replied:
I don't think the plumbing to get the original-operation-version passed all the way down to validation is going to fly.  It's almost as ugly as hacking in versioned validation.  This info is lost deep deep in the bowels of the conversion logic.

I see only bad options here:

1) don't rename
2) return errors with the wrong name in some cases (one way or the other)
3) return errors with both names
4) block this on getting real plumbing laid

Every time I explain that `backend` is the default backend, people ask me why it's not called `default Backend`.  I lean towards (2) or (3).

What if we said "defaultBackend (aka backend in v1beta1)" here and left a big comment blaming @lavalamp for not giving us versioned validation?

liggitt commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -209,10 +209,10 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
+		allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child("defaultBackend"))...)
 	} else if len(spec.Rules) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `backend` or `rules` must be specified"))
+		allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `defaultBackend` or `rules` must be specified"))
```
same comment here about confusing errors for v1beta1 API clients

thockin commented on staging/src/k8s.io/api/networking/v1/types.go:
```diff
@@ -201,3 +201,172 @@ type NetworkPolicyList struct {
+	// The default backend capable of servicing requests that don't match any
+	// rule. At least one of 'defaultBackend' or 'rules' must be specified. This field
+	// is optional to allow the loadbalancer controller or defaulting logic to
+	// specify a global default.
```
Should we more strongly encourage this to be populated?  We should at least say something like "If this is not specified, any request which does not match a rule will be handled in an implementation-defined way (e.g. a generic 404 or a 5xx)."

cmluciano replied:
Not sure since the last comment in here suggests that there was an original purpose for it to be optional. I added your suggested comment.

thockin replied:
I think we can't really make it required because of round-trip, but in practice the global default has turned out to be a DISASTER.  Let's NOT make this mistake again.

@rramkumar1 @aledbf  One thing we might do is throw an event from ingress controllers warning that no defaultBackend is set, and that unhandled URLs may fall through to the global default.

thockin replied:
We dont have a mechanism to feed "warnings" back through API ops (sadly)

liggitt replied:
> We dont have a mechanism to feed "warnings" back through API ops (sadly)

nowhere close to available yet, but I started looking into doing that via warning headers in https://github.com/kubernetes/kubernetes/pull/73032

thockin commented on staging/src/k8s.io/api/networking/v1/types.go:
```diff
@@ -201,3 +201,172 @@ type NetworkPolicyList struct {
+	TLS []IngressTLS `json:"tls,omitempty" protobuf:"bytes,2,rep,name=tls"`
+
+	// A list of host rules used to configure the Ingress. If unspecified, or
+	// no rule matches, all traffic is sent to the default backend.
```
s/all//

thockin reviewed (commented):
I wonder if we should accumulate all the PRs needed to make v1 in v1beta3 and then move it over, or otherwise gate it so we don't release a half-cooked v1 ?

cmluciano commented:
> I wonder if we should accumulate all the PRs needed to make v1 in v1beta3 and then move it over, or otherwise gate it so we don't release a half-cooked v1 ?

@liggitt  mentioned that above as well. Let me flip this around and do that instead.

bowei commented:
@cmluciano -- we'll have to flag gate v1beta2 anyway if it's half complete; I would just flag gate the v1 to avoid having to do unnecessary moving of types around.

liggitt commented:
> @cmluciano -- we'll have to flag gate v1beta2 anyway if it's half complete; I would just flag gate the v1 to avoid having to do unnecessary moving of types around.

the issue is more with the client libraries we ship. we don't want to ship incompatible ingress v1 go types that could get built into an app using 1.16-level client libraries, then used against a 1.X server with the v1 API enabled

liggitt commented:
> I would just flag gate the v1 to avoid having to do unnecessary moving of types around.

I share the concern about churn of moving types around. in other API promotions involving changes, we've accumulated each change as a commit and held the PR until all changes were ready before merging

bowei commented:
ahh -- got it.

cmluciano commented:
@liggitt mentioned that a feature branch might be a good option so that we can work in parallel. WDYT @bowei

bowei commented:
Actually that makes a lot of sense. what is the procedure around it?

liggitt commented:
we've used them for a couple projects, server-side apply and rate-limiting. @lavalamp / @apelisse might have pointers to how those were set up.

cmluciano commented:
I made most of the changes but am still a little confused on what we should do with the validation code given that it checks against the internal version as Tim noted. 

We also should get this based on a different branch so that we can work on everything in parallel and not rush out the v1 if we miss a few parts before code freeze.

thockin commented:
I'm not disagreeing.  I am violently agreeing.  Even with a map, I don't
know which version the user was engaging, unless we plumb that all the way
down.

We could make the error be gross and say "defaultBackend (AKA backend in
v1beta1) .."

On Fri, Aug 16, 2019 at 11:34 AM Jordan Liggitt <notifications@github.com>
wrote:

> *@liggitt* commented on this pull request.
> ------------------------------
>
> In pkg/apis/networking/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/80179#discussion_r314841778>
> :
>
> > @@ -209,10 +209,10 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
>  func ValidateIngressSpec(spec *networking.IngressSpec, fldPath *field.Path) field.ErrorList {
>  	allErrs := field.ErrorList{}
>  	// TODO: Is a default backend mandatory?
> -	if spec.Backend != nil {
> -		allErrs = append(allErrs, validateIngressBackend(spec.Backend, fldPath.Child("backend"))...)
> +	if spec.DefaultBackend != nil {
> +		allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child("defaultBackend"))...)
>
> In the mean time, I think we should go with the new name in errors.
>
> The alternative is to plumb operated-as-version strings down into
> validation (and updates). Blech.
>
> as a user, it is maddening to get an API error, follow the instructions,
> and not resolve the error (or produce a different "unknown field" error).
> I'd probably plumb a map of internal to external name for this one until
> the per-version validation approach materializes
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/80179?email_source=notifications&email_token=ABKWAVBHRPVU5OZ6IGSJIVTQE3XLNA5CNFSM4ID2VCCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCB2QV3Y#discussion_r314841778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFMTHEXS674W6UBKSTQE3XLNANCNFSM4ID2VCCA>
> .
>

cmluciano commented:
/test pull-kubernetes-integratio

cmluciano commented:
/retest

cmluciano commented:
/retest

smourapina commented:
@cmluciano @thockin Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this PR still on track for 1.17?

smourapina commented:
@cmluciano @thockin @caseydavenport Can you please comment if this is still on track for milestone 1.17, or should it be moved to 1.18, as there has been no recent progress on it?

cmluciano commented:
We are working on merging more features into this PR so that we can add all v1 changes. We will have more info over the next week or two.

josiahbjorgaard commented:
/milestone v1.18

(code freeze is in less than two days)

cmluciano commented:
Thanks @josiahbjorgaard we plan for this to be combined with other PRs for 1.18

thockin commented:
Revisiting this, I still have zero idea how to solve this.

cmluciano commented:
@thockin Jordan shared some examples at KubeCon with me. I am taking a look at how this is done in the admissionregistration package and will pull some examples out of there.

thockin commented:
I spent time over holidays poking at this.  A "real" fix might be long in coming, but @deads2k shared a cute trick.

Take a look at https://github.com/kubernetes/kubernetes/compare/master...thockin:hack-ingress-versioned-validation?expand=1

If we plumb that down, we can use it to choose a name for the error

cmluciano commented:
@thockin Just saw this now, thanks! I believe this is what my commit is missing to fix the tests.

Commit:
ingress: rename backend to defaultBackend for ingress v1

Per the Ingress v1 KEP, we are renaming the backend field in IngressSpec to
clarify its intended use. Appropriate defaults are added to ensure backend is
converted to the new field type.

ingress: validate backend based on schema.GroupVersion

Signed-off-by: Christopher M. Luciano <cmluciano@us.ibm.com>

cmluciano commented:
@liggitt I think I'm pretty close on this one. The last failure is seems to be the storage test. I think the problem might be that it is missing [requestInfo details like I have in strategy](https://github.com/cmluciano/kubernetes/blob/cml/ingdefaultbackend/pkg/registry/networking/ingress/strategy.go#L75-L82).

```
make test WHAT=./pkg/registry/networking/ingress/storage/...
+++ [0203 16:49:40] Running tests without code coverage
I0203 16:49:43.116015   11059 client.go:361] parsed scheme: "endpoint"
I0203 16:49:43.116161   11059 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:84153270101107355820  <nil> 0 <nil>}]
I0203 16:49:43.118499   11059 client.go:361] parsed scheme: "endpoint"
I0203 16:49:43.118579   11059 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:84153270101107355820  <nil> 0 <nil>}]
I0203 16:49:43.119787   11059 client.go:361] parsed scheme: "endpoint"
I0203 16:49:43.119852   11059 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:84153270101107355820  <nil> 0 <nil>}]
--- FAIL: TestCreate (0.09s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x15d075c]

goroutine 145 [running]:
testing.tRunner.func1(0xc000506300)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x1ba77c0, 0x323af60)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/kubernetes/pkg/registry/networking/ingress.ingressStrategy.Validate(0x215d0e0, 0xc0001e3960, 0x21517c0, 0x3281680, 0x219cec0, 0xc0007045d0, 0x2156e20, 0xc000483080, 0x0, 0x0, ...)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/networking/ingress/strategy.go:79 +0x6c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest.BeforeCreate(0x21ba680, 0xc0001544c0, 0x219cec0, 0xc0007045d0, 0x2156e20, 0xc000483080, 0x48, 0x1cb8440)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/create.go:105 +0x2de
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Create(0xc000299b00, 0x219cec0, 0xc0007045d0, 0x2156e20, 0xc000483080, 0x1f1df58, 0xc0004840f0, 0xc00068a700, 0xc00068a400, 0x20, ...)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:338 +0x81
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest.(*Tester).testCreateHasMetadata(0xc0007043f0, 0x2156e20, 0xc000483080)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go:392 +0x1bb
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest.(*Tester).TestCreate(0xc0007043f0, 0x2156e20, 0xc000482c00, 0xc00082dd38, 0xc00082dd28, 0xc00082de40, 0x2, 0x2)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go:148 +0x109
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/testing.(*Tester).TestCreate(0xc000c6e150, 0x2156e20, 0xc000482c00, 0xc00082de40, 0x2, 0x2)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/testing/tester.go:79 +0xb3
k8s.io/kubernetes/pkg/registry/networking/ingress/storage.TestCreate(0xc000506300)
	/home/cmluciano/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/networking/ingress/storage/storage_test.go:139 +0x411
testing.tRunner(0xc000506300, 0x1f1b500)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350
FAIL	k8s.io/kubernetes/pkg/registry/networking/ingress/storage	0.131s
FAIL
Makefile:185: recipe for target 'test' failed
make: *** [test] Error 1
```

liggitt commented:
I'd recommend something like this in ValidateUpdate and Validate:
```go
var requestGV schema.GroupVersion
if requestInfo, ok := request.RequestInfoFrom(ctx); ok {
	requestGV = schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
}
```

liggitt commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -204,13 +207,17 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
+		}
+		if requestGV == networkingv1beta1.SchemeGroupVersion {
+			allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `backend` or `rules` must be specified"))
+		}
```
this is fragile, and means if we add another version we have to remember to revisit this or we get unvalidated fields!

assuming we are better at naming this field now than we were four years ago, something like this that assumes future versions will continue using the v1 field name might make more sense:
```go
var defaultBackendFieldName string
switch requestGV {
case networkingv1beta1.SchemeGroupVersion, extensionsv1beta1.SchemeGroupVersion:
	defaultBackendFieldName = "backend"
default:
	defaultBackendFieldName = "defaultBackend"
}

allErrs := field.ErrorList{}
if spec.DefaultBackend != nil {
	allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child(defaultBackendFieldName))...)
} else if len(spec.Rules) == 0 {
	allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, fmt.Sprintf("either `%s` or `rules` must be specified", defaultBackendFieldName)))
}
```

cmluciano replied:
When I tried it this way it doesn't seem to trigger the validation error that is expected in strategy_test.go. I'm scouring the relevant files under pkg/registry/networking/ingress/* to see if I am accidentally using the old backend type but I have not found a mistake yet.

cmluciano replied:
Nvm copy/paste error, I found my mistake

cmluciano replied:
Actually even after fixing an indenting problem, this still doesnt work because it passes only a string to the ingressbackend validation. I kept the if/else for now.

liggitt replied:
also, this code doesn't currently handle extensions/v1beta1

cmluciano commented:
/retest

robscott commented:
/assign

cmluciano commented:
/retest

liggitt commented:
closing in favor of https://github.com/kubernetes/kubernetes/pull/88041

/close
